File: Examples/ObjcDemo/RixSDKPodDemo/RixEngineObjcDemo/Base/BaseMenuVC.m
#import "BaseMenuVC.h"

@interface MenuCardCell ()

@property (nonatomic, strong) UIView *cardView;
@property (nonatomic, strong) UILabel *titleLabel;
@property (nonatomic, strong) UILabel *subtitleLabel;
@property (nonatomic, strong) UIImageView *chevronImageView;

@end

@implementation MenuCardCell

+ (NSString *)reuseID {
    return @"MenuCardCell";
}

- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier {
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    if (self) {
        [self setupUI];
    }
    return self;
}

- (void)setHighlighted:(BOOL)highlighted animated:(BOOL)animated {
    [super setHighlighted:highlighted animated:animated];
    CGFloat scale = highlighted ? 0.985 : 1.0;
    CGFloat alpha = highlighted ? 0.82 : 1.0;
    [UIView animateWithDuration:animated ? 0.15 : 0 animations:^{
        self.cardView.transform = CGAffineTransformMakeScale(scale, scale);
        self.cardView.alpha = alpha;
    }];
}

- (void)configureWithTitle:(NSString *)title subtitle:(NSString *)subtitle {
    self.titleLabel.text = title;
    self.subtitleLabel.text = subtitle;
}

- (void)setupUI {
    self.selectionStyle = UITableViewCellSelectionStyleNone;
    self.backgroundColor = [UIColor clearColor];
    self.contentView.backgroundColor = [UIColor clearColor];

    self.cardView = [[UIView alloc] init];
    self.cardView.translatesAutoresizingMaskIntoConstraints = NO;
    self.cardView.backgroundColor = [UIColor whiteColor];
    self.cardView.layer.cornerRadius = 18;

    self.titleLabel = [[UILabel alloc] init];
    self.titleLabel.translatesAutoresizingMaskIntoConstraints = NO;
    self.titleLabel.font = [UIFont systemFontOfSize:18 weight:UIFontWeightSemibold];
    self.titleLabel.textColor = [UIColor colorWithRed:0.12 green:0.13 blue:0.20 alpha:1];
    self.titleLabel.numberOfLines = 1;

    self.subtitleLabel = [[UILabel alloc] init];
    self.subtitleLabel.translatesAutoresizingMaskIntoConstraints = NO;
    self.subtitleLabel.font = [UIFont systemFontOfSize:13 weight:UIFontWeightRegular];
    self.subtitleLabel.textColor = [UIColor colorWithRed:0.56 green:0.57 blue:0.64 alpha:1];
    self.subtitleLabel.numberOfLines = 2;

    self.chevronImageView = [[UIImageView alloc] init];
    self.chevronImageView.translatesAutoresizingMaskIntoConstraints = NO;
    if (@available(iOS 13.0, *)) {
        self.chevronImageView.image = [UIImage systemImageNamed:@"chevron.right"];
    }
    self.chevronImageView.contentMode = UIViewContentModeScaleAspectFit;
    self.chevronImageView.tintColor = [UIColor colorWithRed:0.78 green:0.79 blue:0.84 alpha:1];

    [self.contentView addSubview:self.cardView];
    [self.cardView addSubview:self.titleLabel];
    [self.cardView addSubview:self.subtitleLabel];
    [self.cardView addSubview:self.chevronImageView];

    [NSLayoutConstraint activateConstraints:@[
        [self.cardView.leadingAnchor constraintEqualToAnchor:self.contentView.leadingAnchor constant:20],
        [self.cardView.trailingAnchor constraintEqualToAnchor:self.contentView.trailingAnchor constant:-20],
        [self.cardView.topAnchor constraintEqualToAnchor:self.contentView.topAnchor constant:6],
        [self.cardView.bottomAnchor constraintEqualToAnchor:self.contentView.bottomAnchor constant:-6],

        [self.titleLabel.leadingAnchor constraintEqualToAnchor:self.cardView.leadingAnchor constant:16],
        [self.titleLabel.topAnchor constraintEqualToAnchor:self.cardView.topAnchor constant:14],
        [self.titleLabel.trailingAnchor constraintLessThanOrEqualToAnchor:self.chevronImageView.leadingAnchor constant:-12],

        [self.subtitleLabel.leadingAnchor constraintEqualToAnchor:self.titleLabel.leadingAnchor],
        [self.subtitleLabel.trailingAnchor constraintLessThanOrEqualToAnchor:self.chevronImageView.leadingAnchor constant:-12],
        [self.subtitleLabel.topAnchor constraintEqualToAnchor:self.titleLabel.bottomAnchor constant:6],
        [self.subtitleLabel.bottomAnchor constraintEqualToAnchor:self.cardView.bottomAnchor constant:-14],

        [self.chevronImageView.trailingAnchor constraintEqualToAnchor:self.cardView.trailingAnchor constant:-16],
        [self.chevronImageView.centerYAnchor constraintEqualToAnchor:self.cardView.centerYAnchor],
        [self.chevronImageView.widthAnchor constraintEqualToConstant:14],
        [self.chevronImageView.heightAnchor constraintEqualToConstant:14]
    ]];
}

@end

@implementation BaseMenuVC

- (void)viewDidLoad {
    [super viewDidLoad];
    
    if ([self menuAppearance] == MenuAppearanceCard) {
        self.view.backgroundColor = [UIColor colorWithRed:0.95 green:0.95 blue:0.98 alpha:1.0];
    } else {
        self.view.backgroundColor = [UIColor whiteColor];
    }
    
    [self setupSDK];
    [self setupTableView];
}

- (void)setupTableView {
    self.tableView = [[UITableView alloc] initWithFrame:CGRectZero style:UITableViewStylePlain];
    self.tableView.translatesAutoresizingMaskIntoConstraints = NO;
    self.tableView.dataSource = self;
    self.tableView.delegate = self;
    [self.tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:@"MenuCell"];
    [self.tableView registerClass:[MenuCardCell class] forCellReuseIdentifier:[MenuCardCell reuseID]];
    self.tableView.tableFooterView = [[UIView alloc] init];
    
    [self applyMenuAppearance];
    [self.view addSubview:self.tableView];
    
    [NSLayoutConstraint activateConstraints:@[
        [self.tableView.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor],
        [self.tableView.leadingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.leadingAnchor],
        [self.tableView.trailingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.trailingAnchor],
        [self.tableView.bottomAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.bottomAnchor]
    ]];
}

- (void)applyMenuAppearance {
    switch ([self menuAppearance]) {
        case MenuAppearancePlain:
            self.tableView.backgroundColor = [UIColor whiteColor];
            self.tableView.separatorStyle = UITableViewCellSeparatorStyleSingleLine;
            self.tableView.showsVerticalScrollIndicator = YES;
            self.tableView.contentInset = UIEdgeInsetsZero;
            self.tableView.estimatedRowHeight = 54;
            self.tableView.rowHeight = 54;
            if (@available(iOS 15.0, *)) {
                self.tableView.sectionHeaderTopPadding = 0;
            }
            break;
        case MenuAppearanceCard:
            self.tableView.backgroundColor = self.view.backgroundColor;
            self.tableView.separatorStyle = UITableViewCellSeparatorStyleNone;
            self.tableView.showsVerticalScrollIndicator = NO;
            self.tableView.contentInset = UIEdgeInsetsMake(10, 0, 18, 0);
            self.tableView.estimatedRowHeight = 92;
            self.tableView.rowHeight = UITableViewAutomaticDimension;
            if (@available(iOS 15.0, *)) {
                self.tableView.sectionHeaderTopPadding = 0;
            }
            break;
    }
}

- (void)setupSDK {}
- (MenuAppearance)menuAppearance { return MenuAppearancePlain; }
- (NSString *)menuSubtitleAtIndex:(NSInteger)index { return nil; }

#pragma mark - UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.menuItems.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    if ([self menuAppearance] == MenuAppearanceCard) {
        MenuCardCell *cell = [tableView dequeueReusableCellWithIdentifier:[MenuCardCell reuseID] forIndexPath:indexPath];
        MenuItem *item = self.menuItems[indexPath.row];
        [cell configureWithTitle:item.title subtitle:[self menuSubtitleAtIndex:indexPath.row] ?: @""];
        return cell;
    } else {
        UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"MenuCell" forIndexPath:indexPath];
        cell.textLabel.text = self.menuItems[indexPath.row].title;
        cell.textLabel.font = [UIFont systemFontOfSize:16 weight:UIFontWeightRegular];
        cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
        return cell;
    }
}

#pragma mark - UITableViewDelegate

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return [self menuAppearance] == MenuAppearanceCard ? UITableViewAutomaticDimension : 54;
}

- (CGFloat)tableView:(UITableView *)tableView estimatedHeightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return [self menuAppearance] == MenuAppearanceCard ? 92 : 54;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    UIViewController *vc = self.menuItems[indexPath.row].makeVC();
    [self.navigationController pushViewController:vc animated:YES];
}

@end

File: SwiftPodDemo/AlxAdsDemo/Base/BaseMenuVC.swift
//
//  BaseMenuVC.swift
//  AlxAdsDemo
//
//  Created by liu weile on 2025/3/31.
//

import UIKit

/// 以 UITableView 列表形式展示菜单入口的基类。
/// 子类只需覆盖 `menuItems` 提供入口数据，以及覆盖 `setupSDK()` 做 SDK 初始化。
class BaseMenuVC: BaseUIViewController {

    typealias MenuItem = (title: String, makeVC: () -> UIViewController)
    enum MenuAppearance {
        case plain
        case card
    }

    // MARK: - Subclass interface

    /// 子类覆盖，返回菜单条目列表
    var menuItems: [MenuItem] { [] }

    /// 子类覆盖，在 viewDidLoad 中做 SDK 初始化（默认空实现）
    func setupSDK() {}

    /// 子类覆盖，控制菜单视觉样式
    var menuAppearance: MenuAppearance { .plain }

    /// 子类覆盖，为卡片样式提供副标题
    func menuSubtitle(at index: Int) -> String? { nil }

    // MARK: - Private

    public let tableView = UITableView(frame: .zero, style: .plain)

    // MARK: - Lifecycle

    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = menuAppearance == .card
            ? UIColor(red: 0.95, green: 0.95, blue: 0.98, alpha: 1.0)
            : .white
        setupSDK()
        setupTableView()
    }

    // MARK: - Layout

    private func setupTableView() {
        tableView.translatesAutoresizingMaskIntoConstraints = false
        tableView.dataSource = self
        tableView.delegate = self
        tableView.register(UITableViewCell.self, forCellReuseIdentifier: "MenuCell")
        tableView.register(MenuCardCell.self, forCellReuseIdentifier: MenuCardCell.reuseID)
        tableView.tableFooterView = UIView()
        applyMenuAppearance()
        view.addSubview(tableView)

        NSLayoutConstraint.activate([
            tableView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor),
            tableView.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor),
            tableView.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor),
            tableView.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor),
        ])
    }

    private func applyMenuAppearance() {
        switch menuAppearance {
        case .plain:
            tableView.backgroundColor = .white
            tableView.separatorStyle = .singleLine
            tableView.showsVerticalScrollIndicator = true
            tableView.contentInset = .zero
            tableView.estimatedRowHeight = 54
            tableView.rowHeight = 54
            if #available(iOS 15.0, *) {
                tableView.sectionHeaderTopPadding = 0
            }
        case .card:
            tableView.backgroundColor = view.backgroundColor
            tableView.separatorStyle = .none
            tableView.showsVerticalScrollIndicator = false
            tableView.contentInset = UIEdgeInsets(top: 10, left: 0, bottom: 18, right: 0)
            tableView.estimatedRowHeight = 92
            tableView.rowHeight = UITableView.automaticDimension
            if #available(iOS 15.0, *) {
                tableView.sectionHeaderTopPadding = 0
            }
        }
    }
}

// MARK: - UITableViewDataSource
extension BaseMenuVC: UITableViewDataSource {

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        menuItems.count
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        if menuAppearance == .card {
            guard let cell = tableView.dequeueReusableCell(
                withIdentifier: MenuCardCell.reuseID,
                for: indexPath
            ) as? MenuCardCell else {
                return UITableViewCell()
            }
            let item = menuItems[indexPath.row]
            cell.configure(
                title: item.title,
                subtitle: menuSubtitle(at: indexPath.row) ?? ""
            )
            return cell
        } else {
            let cell = tableView.dequeueReusableCell(withIdentifier: "MenuCell", for: indexPath)
            cell.textLabel?.text = menuItems[indexPath.row].title
            cell.textLabel?.font = .systemFont(ofSize: 16, weight: .regular)
            cell.accessoryType = .disclosureIndicator
            return cell
        }
    }
}

// MARK: - UITableViewDelegate
extension BaseMenuVC: UITableViewDelegate {

    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        menuAppearance == .card ? UITableView.automaticDimension : 54
    }

    func tableView(_ tableView: UITableView, estimatedHeightForRowAt indexPath: IndexPath) -> CGFloat {
        menuAppearance == .card ? 92 : 54
    }

    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        let vc = menuItems[indexPath.row].makeVC()
        navigationController?.pushViewController(vc, animated: true)
    }
}

final class MenuCardCell: UITableViewCell {
    static let reuseID = "MenuCardCell"

    private let cardView = UIView()
    private let titleLabel = UILabel()
    private let subtitleLabel = UILabel()
    private let chevronImageView = UIImageView()

    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        setupUI()
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func setHighlighted(_ highlighted: Bool, animated: Bool) {
        super.setHighlighted(highlighted, animated: animated)
        let scale: CGFloat = highlighted ? 0.985 : 1.0
        let alpha: CGFloat = highlighted ? 0.82 : 1.0
        UIView.animate(withDuration: animated ? 0.15 : 0) {
            self.cardView.transform = CGAffineTransform(scaleX: scale, y: scale)
            self.cardView.alpha = alpha
        }
    }

    func configure(title: String, subtitle: String) {
        titleLabel.text = title
        subtitleLabel.text = subtitle
    }

    private func setupUI() {
        selectionStyle = .none
        backgroundColor = .clear
        contentView.backgroundColor = .clear

        cardView.translatesAutoresizingMaskIntoConstraints = false
        cardView.backgroundColor = .white
        cardView.layer.cornerRadius = 18

        titleLabel.translatesAutoresizingMaskIntoConstraints = false
        titleLabel.font = .systemFont(ofSize: 18, weight: .semibold)
        titleLabel.textColor = UIColor(red: 0.12, green: 0.13, blue: 0.20, alpha: 1)
        titleLabel.numberOfLines = 1

        subtitleLabel.translatesAutoresizingMaskIntoConstraints = false
        subtitleLabel.font = .systemFont(ofSize: 13, weight: .regular)
        subtitleLabel.textColor = UIColor(red: 0.56, green: 0.57, blue: 0.64, alpha: 1)
        subtitleLabel.numberOfLines = 2

        chevronImageView.translatesAutoresizingMaskIntoConstraints = false
        chevronImageView.image = UIImage(systemName: "chevron.right")
        chevronImageView.contentMode = .scaleAspectFit
        chevronImageView.tintColor = UIColor(red: 0.78, green: 0.79, blue: 0.84, alpha: 1)

        contentView.addSubview(cardView)
        cardView.addSubview(titleLabel)
        cardView.addSubview(subtitleLabel)
        cardView.addSubview(chevronImageView)

        NSLayoutConstraint.activate([
            cardView.leadingAnchor.constraint(equalTo: contentView.leadingAnchor, constant: 20),
            cardView.trailingAnchor.constraint(equalTo: contentView.trailingAnchor, constant: -20),
            cardView.topAnchor.constraint(equalTo: contentView.topAnchor, constant: 6),
            cardView.bottomAnchor.constraint(equalTo: contentView.bottomAnchor, constant: -6),

            titleLabel.leadingAnchor.constraint(equalTo: cardView.leadingAnchor, constant: 16),
            titleLabel.topAnchor.constraint(equalTo: cardView.topAnchor, constant: 14),
            titleLabel.trailingAnchor.constraint(lessThanOrEqualTo: chevronImageView.leadingAnchor, constant: -12),

            subtitleLabel.leadingAnchor.constraint(equalTo: titleLabel.leadingAnchor),
            subtitleLabel.trailingAnchor.constraint(lessThanOrEqualTo: chevronImageView.leadingAnchor, constant: -12),
            subtitleLabel.topAnchor.constraint(equalTo: titleLabel.bottomAnchor, constant: 6),
            subtitleLabel.bottomAnchor.constraint(equalTo: cardView.bottomAnchor, constant: -14),

            chevronImageView.trailingAnchor.constraint(equalTo: cardView.trailingAnchor, constant: -16),
            chevronImageView.centerYAnchor.constraint(equalTo: cardView.centerYAnchor),
            chevronImageView.widthAnchor.constraint(equalToConstant: 14),
            chevronImageView.heightAnchor.constraint(equalToConstant: 14)
        ])
    }
}
